rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy

```diff
@@ -71,30 +71,39 @@ abstract class CloudProviderBakeHandler {
    * bake.
    */
   String produceBakeKey(String region, BakeRequest bakeRequest) {
-    // TODO(duftler): Consider whether template_file_name, extended_attributes and var_file_name should be
-    // taken into account when composing the bake key.
     bakeRequest.with {
-      def bakeKey = "bake:$cloud_provider_type:$base_os"
+      def keys = [
+              'bake',
+              cloud_provider_type,
+              base_os
+      ]
 
       if (base_ami) {
-        bakeKey += ":$base_ami"
+        keys << base_ami
```

Comment:
Now that there are so many optional key segments, maybe it makes sense to extract an `addIfPresent()` utility method?